docs/Hardware-Agnostic-Training.md

```diff
@@ -0,0 +1,67 @@
+
+We have comprehensive documentation on [getting started with training](Getting-Started-Training) in levanter, the [configuration file](Configuration-Guide.md), and [training on your own dataset](Training-On-Your-Data.md) that you should check out for more details. We also have a notebook tutorial on [how to add FSDP to custom architectures](https://colab.research.google.com/drive/1QX4yH3zRFF3Xiibf1aahETcSQ5nbcUMz) implemented in Levanter with Haliax.
+
+
+### How the Model Checkpoint is Saved
+The frequency of checkpoint saves can be adjusted by configuring the `checkpointer.save_interval` parameter in your training command. This parameter requires an integer value, which specifies the interval (in minutes) between each checkpoint save. By default, checkpoints are saved every 15 minutes.
+
+Whenever a checkpoint is saved, Levanter creates a new directory named after the training run ID within the path specified by `trainer.checkpointer.base_path`. This directory stores each checkpoint file. Every checkpoint file containes:
+
+* the model's parameters
+* the current optimizer and trainer states
+* meta data on the current training step and the time the training step was saved
+
+```
+user@DN0a2335e6 rj96a79n/step-1591 % ls
+metadata.json	model		training_state	worker-0.cert
+```
+
+### Moving the Checkpoint File to Google Storage
+To switch to a new hardware configuration mid training run, you need to move your checkpoint file and your training data over to the new hardware environemnt. If you started your training run on a GPU server and want to move it to a TPU VM instances, you can use the following command to upload your checkpoint file to a google storage drive that your TPU VM can access.
+
+```bash
+gsutil -m cp -r <GPU MODEL CHECKPT PATH> gs://<SOMEWHERE>
+```
+You need to be logged into your GCP account in your current environment for this command to run properly. Information on how to set this up can be found [here](Getting-Started-TPU-VM.md).
+
+### Re-Configuring `per_device_parallelism`
+When you switch to a new hardware setup, whether its from GPU to TPU or simply agjusting the number of GPUs, it is import you keep your training batch size (`trainer.train_batch_size`) the same. You should however, adjust your `trainer.per_device_parallelism` size to match the new number of accelorators you will be using. For example, if you are moving to a v3-32 TPU VM instances, this means you are using v3 TPUs with 32 TensorCores (I found [this page](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm) helpful for understanding TPU terminology). Each tensor core can handle a portion of your total batch, so you might set your `trainer.per_device_parallelism` size to be 256 / 32 = 8. If you were to set `trainer.per_device_parallelism` to be some suitable value smaller than 8 (like 1, 2, or 4), then Levanter would perform gradient accumlation until 8 examples have been processed by the model on this accelorator.
```

Comment:
```suggestion
When you switch to a new hardware setup, whether its from GPU to TPU or simply adjusting the number of GPUs, it is important you keep your training batch size (`trainer.train_batch_size`) the same. You should however, adjust your `trainer.per_device_parallelism` size to match the new number of accelerators you will be using. For example, if you are moving to a v3-32 TPU VM instances, this means you are using v3 TPUs with 32 TensorCores (I found [this page](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm) helpful for understanding TPU terminology). Each tensor core can handle a portion of your total batch, so you might set your `trainer.per_device_parallelism` size to be 256 / 32 = 8. If you were to set `trainer.per_device_parallelism` to be some suitable value smaller than 8 (like 1, 2, or 4), then Levanter would perform gradient accumulation until 8 examples have been processed by the model on this accelerator.
```

Reply:
I'm not totally sure it is important to keep the train batch size the same, but it's probably a good idea. changing the number of train steps is probably asking for trouble b/c it messes with the LR schedule